### ISSUE TITLE
backpropagation manually not equal with weights keras

### ISSUE THREAD

**Issue by ehsanakefi** [keras-team/keras#18324]
I calculated the backpropagation manually and also obtained the weights with the keras, but it is not equal. [calculated the backpropagation manually picture](https://i.stack.imgur.com/5pqhP.png) I calculated the backpropagation manually is 0.19926 and output keras weights is 0.19882499
[this is my mlp picture](https://i.stack.imgur.com/qQSZZ.png)
```
from keras.mixed_precision.loss_scale_optimizer import optimizer 
import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

xtrain=[[0,0]]
ytrain=[[0]]
inputs = keras.Input(shape=(2,))
x=keras.layers.Dense(5,activation="sigmoid", use_bias=False, bias_initializer='zeros')(inputs)
output=layers.Dense(1,activation="sigmoid", use_bias=False, bias_initializer='zeros')(x)
model=keras.Model(inputs=inputs,outputs=output,name="xor")
model.compile(loss='mean_absolute_error',
              optimizer=tf.keras.optimizers.SGD(
    learning_rate=0.01,
    momentum=0.0,
    nesterov=False,
    weight_decay=0.0,
    clipnorm=None,
    clipvalue=None,
    global_clipnorm=None,
    use_ema=False,
    ema_momentum= 0,
    ema_overwrite_frequency=None,
    jit_compile=True,
    name="SGD",

))
i=0
for layerNum, layer in enumerate(model.layers):
    if(i==1):
      layer.set_weights([np.array([[0.1,0.1,0.1,0.1,0.1],[0.1,0.1,0.1,0.1,0.1]])])
    if(i==2):
      layer.set_weights([np.array([[0.2],[0.2],[0.2],[0.2],[0.2]])])
    i=i+1

history=model.fit(xtrain,ytrain,batch_size=1,epochs=1)

for layerNum, layer in enumerate(model.layers):

    weights = layer.get_weights()
    print("weights=",weights)
    print(layerNum,"layerNum") 
```

**Comment by grasskin** [comment]
I think your issue stems from a mistake with the loss function gradient. You are using mean absolute error as your loss which has a gradient of 1 or -1 with respect to its input, more explicitly: `abs(ytrain-model(xtrain))`.

Let's focus on calculating the new value for the weight that you've highlighted. Let us call this $w$, let our loss be $L$ and let's stick to $a$ as the output of our activation function (sigmoid). $o$ will be the output before applying activation. I avoid subscripts for clarity given we are calculating everything for a single weight.

We are interested in finding the gradient of our loss with respect to our weight $\frac{\partial L}{\partial w}$. By chain rule:

$$\frac{\partial L}{w}=\frac{\partial L}{\partial a}\frac{\partial a}{\partial o}\frac{\partial o}{\partial w}$$

You correctly calculated the second and third terms $$\frac{\partial o}{\partial a}=o(1-o)=0.235004$$

Which is the derivative of the sigmoid (itself times 1 - itself).

And $$\frac{\partial o}{\partial w}=\frac{w*0.5+...}{\partial w}=0.5$$

Your mistake is that $$\frac{\partial L}{\partial a}=1\neq 0.622459$$

The derivative of our mean absolute error is either 1 or -1, not the final error.

Pluging this all in to our update rule $$w'=w-0.01*\frac{\partial L}{w}=0.2-0.01 * 0.5 * 0.235004=0.198825$$

Which is what Keras outputs. A couple helpful tips for anyone working with gradients by hand:

1. You can get intermediate outputs in Keras very easily, create a second model that takes in the same input layer and a list of all your model outputs as outputs. Specifically, given your `model`, `inputs` as an input layer and `data` as your training data:
```python
activation_outputs = [layer.output for layer in model.layers]
activation_model = tf.keras.Model(inputs, activation_outputs)
activation_model(data)
```

will output
```
[<tf.Tensor: shape=(1, 2), dtype=float32, numpy=array([[0., 0.]], dtype=float32)>,
 <tf.Tensor: shape=(1, 5), dtype=float32, numpy=array([[0.5, 0.5, 0.5, 0.5, 0.5]], dtype=float32)>,
 <tf.Tensor: shape=(1, 1), dtype=float32, numpy=array([[0.62245935]], dtype=float32)>]
```
2. Tensorflow `GradientTape` is a powerful way to get intermediate gradients. For example:

```python
with tf.GradientTape() as tape:
    out = model(data)
    loss = keras.losses.mean_absolute_error(0, out)
tape.gradient(loss,model.trainable_variables)
```

will output
```
[<tf.Tensor: shape=(2, 5), dtype=float32, numpy=
 array([[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]], dtype=float32)>,
 <tf.Tensor: shape=(5, 1), dtype=float32, numpy=
 array([[0.11750185],
        [0.11750185],
        [0.11750185],
        [0.11750185],
        [0.11750185]], dtype=float32)>]
```

giving you more information on gradients.
